Add Integration Tests For BrandingService

mark-monteiro opened the pull request:
Adds support for integration tests with a sample implementation of tests for the endpoints in `BrandingService`.

**Issues**
Creating integration tests to validate the existing behavior of endpoints should be mandatory before migrating the endpoint to the ASP framework in #2872.

mark-monteiro commented on .gitignore:
```diff
@@ -40,6 +40,7 @@ CorePlugins*/
 ProgramData-Server*/
 ProgramData-UI*/
 MediaBrowser.WebDashboard/jellyfin-web/**
+tests/**/launchSettings.json
```
This file is generated every time the solution is opened with Visual Studio but it is not a required file for an integration tests project so I think it's safe to ignore it. The tests run fine without this file existing.

mark-monteiro replied:
Figured out why this was being added to the solution and removed this from the .gitignore file

mark-monteiro commented on tests/MediaBrowser.Api.Tests/JellyfinApplicationFactory.cs:
```diff
@@ -0,0 +1,124 @@
+    public class JellyfinApplicationFactory : WebApplicationFactory<Startup>
+    {
+        private static readonly string _testPathRoot = Path.Combine(Path.GetTempPath(), "jellyfin-test-data");
+        private static readonly ConcurrentBag<ApplicationHost> _appHosts = new ConcurrentBag<ApplicationHost>();
```
I wasn't actually sure if a thread-safe collection is necessary here, but I just threw it in to be safe. I think it might be necessary for running tests in parallel. Can switch to a regular collection if anyone can provide insight that shows it isn't needed.

mark-monteiro commented on tests/MediaBrowser.Api.Tests/JellyfinApplicationFactory.cs:
```diff
@@ -0,0 +1,124 @@
+            };
+
+            // Use a temporary directory for the application paths
+            var webHostPathRoot = Path.Combine(_testPathRoot, "test-host-" + Path.GetFileNameWithoutExtension(Path.GetRandomFileName()));
```
Each web host that is created gets its own application paths base directory

JustAMan commented:
@mark-monteiro seems like tests are failing...

mark-monteiro commented:
> @mark-monteiro seems like tests are failing...

Yeah sorry, hit that request for review to soon. The project works from Visual Studio, but doesn't build from the command line, I'm trying to figure out why :/

joshuaboniface commented:
/azp run

Bond-009 commented on tests/MediaBrowser.Api.Tests/JellyfinApplicationFactory.cs:
```diff
@@ -0,0 +1,117 @@
+
+            // Create the logging config file
+            // TODO: We shouldn't need to do this since we are only logging to console
+            Program.InitLoggingConfigFile(appPaths).Wait();
```
Use `GetAwaiter().GetResult()` to await a task in a sync context

Bond-009 commented on tests/MediaBrowser.Api.Tests/JellyfinApplicationFactory.cs:
```diff
@@ -0,0 +1,117 @@
+            // Finish initializing the app host
+            var appHost = (CoreAppHost)testServer.Services.GetRequiredService<IApplicationHost>();
+            appHost.ServiceProvider = testServer.Services;
+            appHost.InitializeServices().Wait();
+            appHost.RunStartupTasksAsync().Wait();
```
Same

Bond-009 commented:
Tests are failing?

mark-monteiro commented:
> Tests are failing?

Yes, I've fixed the exception causing the failures in https://github.com/jellyfin/jellyfin/pull/2989. I'm a bit unsure of the implications of that fix so maybe you could provide some insight there as to why the write lock and connection were shared between the two repositories in the first place.

crobibero commented:
Should analyzers be added to the testing project?

mark-monteiro commented:
> Should analyzers be added to the testing project?

Good catch, I meant to do that and never got around to it. I've added the FxCop and StyleCop analyzers (the Serilog and MultiThreading analyzers did not seem relevant here). This actually caught a bug for an object that wasn't disposed correctly so I've fixed that as well.

I've configured the analyzers with a custom ruleset that inherits the rules from the main solution ruleset, overriding a few rules as appropriate. @Bond-009 You might want to review again to double-check my approach here.

Bond-009 commented on tests/MediaBrowser.Api.Tests/MediaBrowser.Api.Tests.ruleset:
```diff
@@ -0,0 +1,22 @@
+  <Rules AnalyzerId="Microsoft.CodeAnalysis.FxCopAnalyzers" RuleNamespace="Microsoft.Design">
+    <!-- CA1707: Identifiers should not contain underscores -->
+    <Rule Id="CA1707" Action="None" />
+    <!-- CA2007: Consider calling ConfigureAwait on the awaited task -->
```
Why disable this?

mark-monteiro replied:
AFAIK there is no synchronization context used during the tests and this is not library code that will be consumed somewhere else, so it's not necessary

Bond-009 commented on tests/MediaBrowser.Api.Tests/MediaBrowser.Api.Tests.ruleset:
```diff
@@ -0,0 +1,22 @@
+    <Rule Id="CA1707" Action="None" />
+    <!-- CA2007: Consider calling ConfigureAwait on the awaited task -->
+    <Rule Id="CA2007" Action="None" />
+    <!-- CA2234: Pass system uri objects instead of strings -->
```
Doesn't the main ruleset disable this? If not, maybe it should

mark-monteiro replied:
I was thinking that this rule could be useful in the main projects so I just disabled it here

mark-monteiro commented:
> This new ruleset should be shared by all test projects, no?

I've renamed it and moved it up one directory so that the other test projects can reference it as well

JustAMan commented on tests/MediaBrowser.Api.Tests/JellyfinApplicationFactory.cs:
```diff
@@ -101,22 +98,21 @@ protected override TestServer CreateServer(IWebHostBuilder builder)
             var appHost = (CoreAppHost)testServer.Services.GetRequiredService<IApplicationHost>();
             appHost.ServiceProvider = testServer.Services;
-            appHost.InitializeServices();
-            appHost.FindParts();
```
why is this no longer needed?

mark-monteiro replied:
Those weren't removed, I just updated them to run synchronously: https://github.com/jellyfin/jellyfin/pull/2945/files#diff-01322dfde28c35f44a436fc2ba596206R101
